pkg/compactor/compactor.go

```diff
@@ -0,0 +1,282 @@
+	cfg.retryMaxBackoff = time.Minute
+
+	f.Var(
+		&cfg.BlockRanges,
+		"compactor.block-ranges",
+		"Comma separated list of compaction ranges expressed in the time duration format")
+
+	f.DurationVar(
+		&cfg.ConsistencyDelay,
+		"compactor.consistency-delay",
+		30*time.Minute,
+		fmt.Sprintf("Minimum age of fresh (non-compacted) blocks before they are being processed. Malformed blocks older than the maximum of consistency-delay and %s will be removed.", compact.MinimumAgeForRemoval))
+
+	f.IntVar(
+		&cfg.BlockSyncConcurrency,
+		"compactor.block-sync-concurrency",
+		20,
+		"Number of goroutines to use when syncing block metadata from object storage")
+
+	f.StringVar(
+		&cfg.DataDir,
+		"compactor.data-dir",
+		"./data",
+		"Data directory in which to cache blocks and process compactions")
+
+	f.DurationVar(
+		&cfg.CompactionInterval,
+		"compactor.compaction-interval",
+		time.Hour,
+		"The frequency at which the compaction runs")
+
+	f.IntVar(
+		&cfg.CompactionRetries,
+		"compactor.compaction-retries",
+		3,
+		"How many times to retry a failed compaction during a single compaction interval")
+}
+
+// Compactor is a multi-tenant TSDB blocks compactor based on Thanos.
+type Compactor struct {
+	compactorCfg Config
+	storageCfg   cortex_tsdb.Config
+	logger       log.Logger
+
+	// Underlying compactor used to compact TSDB blocks.
+	tsdbCompactor tsdb.Compactor
+
+	// Client used to run operations on the bucket storing blocks.
+	bucketClient objstore.Bucket
+
+	// Wait group used to wait until the internal go routine completes.
+	runner sync.WaitGroup
+
+	// Context used to run compaction and its cancel function to
+	// safely interrupt it on shutdown.
+	ctx       context.Context
+	cancelCtx context.CancelFunc
+
+	// Metrics.
+	compactionRunsStarted   prometheus.Counter
+	compactionRunsCompleted prometheus.Counter
+	compactionRunsFailed    prometheus.Counter
+}
+
+// NewCompactor makes a new Compactor.
+func NewCompactor(compactorCfg Config, storageCfg cortex_tsdb.Config, logger log.Logger, registerer prometheus.Registerer) (*Compactor, error) {
+	ctx, cancelCtx := context.WithCancel(context.Background())
+
+	bucketClient, err := cortex_tsdb.NewBucketClient(ctx, storageCfg, "compactor", logger)
+	if err != nil {
+		cancelCtx()
+		return nil, errors.Wrap(err, "failed to create the bucket client")
+	}
+
+	tsdbCompactor, err := tsdb.NewLeveledCompactor(ctx, registerer, logger, compactorCfg.BlockRanges.ToMillisecond(), downsample.NewPool())
+	if err != nil {
+		cancelCtx()
+		return nil, errors.Wrap(err, "failed to create TSDB compactor")
+	}
+
+	return newCompactor(ctx, cancelCtx, compactorCfg, storageCfg, bucketClient, tsdbCompactor, logger, registerer)
+}
+
+func newCompactor(ctx context.Context, cancelCtx context.CancelFunc, compactorCfg Config, storageCfg cortex_tsdb.Config, bucketClient objstore.Bucket, tsdbCompactor tsdb.Compactor, logger log.Logger, registerer prometheus.Registerer) (*Compactor, error) {
```

Comment:
Would it be more readable on the editor if all arguments are in separate lines?

Reply:
Definitely. Done.